Python/Product/Analysis/Interpreter/Ast/AstAnalysisFunctionWalker.cs

```diff
@@ -111,25 +102,139 @@ FunctionDefinition targetFunction
+            var ce = node.Tests.FirstOrDefault()?.Test as CallExpression;
+            if (ce?.Target is NameExpression ne && ne?.Name == "isinstance" && ce.Args.Count == 2) {
+                var name = (ce.Args[0].Expression as NameExpression)?.Name;
+                var typeName = (ce.Args[1].Expression as NameExpression)?.Name;
+                if (name != null && typeName != null) {
+                    var typeId = BuiltinTypeIdFromName(typeName);
+                    if (typeId != BuiltinTypeId.Unknown) {
+                        _scope.SetInScope(name, new AstPythonConstant(new AstPythonBuiltinType(typeName, typeId)));
+                    }
+                }
+            }
             return base.Walk(node);
         }
 
         public override bool Walk(ReturnStatement node) {
-            foreach (var type in _scope.GetTypesFromValue(_scope.GetValueFromExpression(node.Expression))) {
-                _returnTypes.Add(type);
+            var typesAdded = false;
+            if (node.Expression is NameExpression nex && nex.Name == "self") {
+                // For self return the actual class without added private members
+                var klass = _scope.LookupNameInScopes("__class__", NameLookupContext.LookupOptions.Nonlocal);
+                if (klass != null) {
+                    _returnTypes.Add(klass as IPythonType);
+                    typesAdded = true;
+                }
+            }
+
+            if (!typesAdded) {
+                foreach (var type in _scope.GetTypesFromValue(_scope.GetValueFromExpression(node.Expression))) {
+                    _returnTypes.Add(type);
+                }
+            }
+            return true; // We want to evaluate all code so all private variables in __new__ get defined
+        }
+
+        private IMember GetClass() {
+            bool classmethod, staticmethod;
+            GetMethodType(_target, out classmethod, out staticmethod);
+            var klass = _scope.LookupNameInScopes("__class__", NameLookupContext.LookupOptions.Local);
+            if (!staticmethod && !classmethod) {
+                var cls = klass as IPythonType;
+                if (cls == null) {
+                    klass = null;
+                } else {
+                    klass = new AstPythonConstant(cls, ((cls as ILocatedMember)?.Locations).MaybeEnumerate().ToArray());
+                }
+            }
+            return klass;
+        }
+
+        private AstPythonType GetClassType(IMember klass) {
+            var cls = (klass as AstPythonConstant)?.Type as AstPythonType;
+            if (cls != null) {
+                var self = _scope.LookupNameInScopes("self", NameLookupContext.LookupOptions.Local);
+                if (self == null) {
+                    // Clone type since function analysis can add members that should not be
+                    // visible to the user such as private variables backing public properties.
```

Comment:
Functions can also add members that _should_ be visible though... do we need to add more tricks for hiding things? (If they're totally internal, we can make it an illegal name and filter it out later - pretty sure I used a space prefix elsewhere for this)

Reply:
No, that's a bad idea. We need to show all the names that users may want to access (filtering them from the final completion list is different).

Are we worried about names in the code? Or names that we internally generate?

Reply:
Since this is just for library code, we could also scope not adding members to constructors only (init and new). This is where backing variables are typically created.

Reply:
Right, except it's arbitrary library code. As well as code referenced through search paths rather than in the workspace. I think we should include everything rather than filtering this early.